File: csharp/largest-series-product/LargestSeriesProduct.cs
using System;
using System.Linq;

public static class LargestSeriesProduct
{
    public static long GetLargestProduct(string digits, int span)
    {
        if (span < 0 || span > digits.Length)
        {
            throw new ArgumentException("Span must be between 1 and number of input digits");
        }

        if (!digits.All(char.IsDigit))
        {
            throw new ArgumentException("Digits contain invalid characters");
        }

        if (digits.Length < 1)
        {
            return 1;
        }

        return Enumerable
            .Range(0, digits.Length - span + 1)
            .Select(idx => digits.Substring(idx, span))
            .Max(SeriesProduct);
    }

    private static long SeriesProduct(string series) => series
        .Select(char.GetNumericValue)
        .Select(Convert.ToInt64)
        .Aggregate(1L, (a, b) => a * b);
}

File: kotlin/largest-series-product/src/main/kotlin/LargestSeriesProduct.kt
class Series(private val digits: String) {

    init {
        require(digits.all { it.isDigit() })
    }

    fun getLargestProduct(span: Int): Long {
        require(span > 0 && span <= digits.length)

        return digits
                .windowed(span, 1)
                .maxOf { it.fold(1L) { acc, d -> acc * (d - '0') } }
    }
}
